[JS] Topic: Introduce JavaScript's usage in the Web Browser

subhendudash02 opened the pull request:
<hr>

## Description 📜

Added the documentation in the form of a markdown file.
Added the sub-folders for the screenshots of the output and errors, HTML files and JS files separately.
Fixes #71

<hr>

## Type of change 📝

<!----Please delete options that are not relevant.And in order to tick the check box just but x inside them for example [x] like this----->

- [ ] Audio (Should be in mp3 format Includes speech clarity, Concise ,Low distortion)
- [ ] Video (Animations, screen-recordings, presentations and regular explanatory films are all possibilties etc)
- [x] Documentation (Content Creation in the form of codes or tutorials)
- [ ] Other (If you choose other, Please mention changes below) 

<hr>

## Domain of Contribution 📊

<!----Please delete options that are not relevant.And in order to tick the check box just but x inside them for example [x] like this----->

- [ ] Android Dev (Flutter, Java, Kotlin)
- [ ] Backend Dev (Java, .NET, Python)
- [ ] C/CPP
- [ ] Competitive Programming
- [ ] Cyber Security
- [ ] DSA
- [ ] Database
- [ ] Datascience with Python
- [ ] Datascience with R
- [ ] Frontend Dev (HTML, CSS, JS)
- [ ] Frontend Dev (React, Angular, Vue)
- [ ] Golang
- [ ] Interview Prep
- [ ] Java
- [x] JavaScript
- [ ] MERN
- [ ] Machine Learning
- [ ] Open Source
- [ ] Python
- [ ] Rust
- [ ] Statistics
- [ ] UI/UX

<hr>
 
## Checklist ✅

<!----Please delete options that are not relevant.And in order to tick the check box just but x inside them for example [x] like this----->

- [x] I follow [Contributing Guidelines](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CONTRIBUTING.md) & [Code of conduct](https://github.com/girlscript/winter-of-contributing/blob/main/.github/CODE_OF_CONDUCT.md) of this project.
- [x] I have performed a self-review of my own code or work.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generates no new warnings.
- [x] I'm GWOC'21 contributor

<hr>

<!----Please delete options that are not relevant. And in order to tick the check box just but x inside them for example [x] like this----->

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,124 @@
+
+# Use of `<script>` tag
+
+Externally, JavaScript may be used by generating a separate file with the `.js` extension and linking to the HTML file. This is called *external JavaScript*. Or it can be used inside HTML files which is called *internal JavaScript*.
```
```suggestion
Externally, JavaScript can be used by generating a separate file with the `.js` extension and linking to the HTML file. This is called *external JavaScript*. Or it can be used inside HTML files which is called *internal JavaScript*.
```

Creatoon commented on JavaScript/Topics/3.Web_Browser/HTML/1.External_Js.html:
```diff
@@ -0,0 +1,14 @@
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Document</title>
+
+    <script src = "../JS/script.js"></script>
```
Insert the script tag just before the closing of the body tag, not saying its incorrect but thik of the js file contains that code which will alter the html because it is present in the head tag, it will load first before body and unable to access DOM elements and it will throw an error, that's why move it to before closing of the body tag.

Also it will be a plus point if you write about adding js file in head tag and at the end of the body tag, write about differences

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,124 @@
+
+# Why should JavaScript be chosen as the sole language for front-end web scripting? 
+
+1. JavaScript offers browser functionality as well as a graphical user interface, which HTML and CSS cannot accomplish.
```
```suggestion
1. JavaScript offers browser API as well as a graphical user interface, which HTML and CSS cannot accomplish.
```

Creatoon commented:
@subhendudash02 please also add how js file is compiled inside browser, Talk about JIT compilation, because I think this file should contain how js runs actually

@paramsiddharth can we do this here or we have to create another file to talk about JIT and Js Engine and browser API, because it's very imp concept which is not known to many js developers.

subhendudash02 commented:
@Creatoon Thank you very much for the review! I will do the modifications needed and I will look for JIT compilation when @paramsiddharth approves!

s4-2 commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,176 @@
+## Advantages of JIT compiler
+
+1. Suited for dynamic programming languages.
+2. More secure
```
More secure.

s4-2 commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,176 @@
+4. Less memory usage.
+
+## Disadvantages of JIT compiler
+1. More cache memory.
```
It looks more descriptive if you add
 1)  JIT reserve more cache memory.

subhendudash02 replied:
Oh, that's a mistake, I will correct it now, so it comes under advantages right?

subhendudash02 commented:
@s4-2, @Creatoon and @paramsiddharth I have modified the documentation as suggested! Kindly have a look if there is some issue or extra information I need to add.
Thank you!

s4-2 reviewed (approved):
Looks good for me. Keep it up:)

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,176 @@
+## Where Javascript is used in Front-End?
+
+The structure of a web page is provided by HTML, the layout and rudimentary animations are provided by CSS, and the website's interactivity is provided by JavaScript.
+
```
```suggestion

So, In a nutshell we can say :- HTML = NOUN, CSS = ADJECTIVE, JS = VERB

```

Please add this line, if you find it correct

subhendudash02 replied:
Oh, this looks short and precise! I will add it.

Creatoon replied:
if you solve the solns mentioned previous comments, then there is a button saying "RESOLVE" this, minimize these comments by clicking it

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,176 @@
+
+<br><hr><br>
+
+# How JavaScript files get compiled?
```
```suggestion
# How JavaScript code get compiled?
```
compilation happens for code, not for files

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,176 @@
+
+# How JavaScript files get compiled?
+
+In most low-level languages, the files get compiled and then executed accordingly. But in JavaScript, compilation takes place during the execution of the file. This type of compiler is known as Just-in-time(JIT) compiler.
```
```suggestion
In most low-level languages, the code gets compiled and then executed accordingly. But in JavaScript, compilation takes place during the execution of the file. This type of compilation is known as Just-In-Time(JIT) compilation.
```

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,176 @@
+
+The Source code of JS file is converted into the *byte code*, which is the intermediate conversion.<br> 
+This bytecode gets interpreted by Virtual Machine. Then it is converted to machine understandable code in section-wise (can be per function or per line) dynamically so that the program can run faster.
+
```
Add image to show JIT.
https://duckduckgo.com/?t=lm&q=JIT+Compilation+javascript&atb=v289-4gy&iax=images&ia=images&iai=https%3A%2F%2Fmiro.medium.com%2Fmax%2F1200%2F1*SezjSMIo5Zub4SyCfZ9iKA.png

If you find some other nice picture then neglect mine and add yours

subhendudash02 replied:
Should I include the source of this image?

Creatoon replied:
yes

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,176 @@
+3. JavaScript contains a large library of libraries, such as jQuery, AJAX, and others, that are useful for complex work/projects.
+4. Many front-end engineers prefer JavaScript because of its clear and straightforward syntax.
+5. JavaScript is rapidly improving as a result of its popularity, and its capacity to make web pages more interactive is growing by the day.
+   
```
Also write javascript provides console Api, not in my way, after reading about it

Creatoon commented:
@subhendudash02 Please look at these changes or additions and after that i will approve from my side

subhendudash02 commented:
@Creatoon I have modified the mistakes and added the flowchart. Kindly let me know if it is required to write the source of the image I included.
Thank you!

subhendudash02 commented:
Added the source of the image.

paramsiddharth commented:
@Creatoon @dakshsethi Please review further.

dakshsethi reviewed (approved):
Looks good to me!! 🔥🔥🔥

Creatoon commented on JavaScript/Topics/3.Web_Browser/HTML/2.index.html:
```diff
@@ -0,0 +1,18 @@
+    <script>
+        "use strict";
+        let x = document.getElementById("main");
+        x.innerHTML = "Internal JavaScript";
```
```suggestion
        x.textContent = "Internal JavaScript";
```
innerHTML is also correct but why to use it if we are not setting an HTML element, for text we should textContent, always use right tool for the a specific work

Creatoon commented on JavaScript/Topics/3.Web_Browser/HTML/3.Internal_Js.html:
```diff
@@ -0,0 +1,20 @@
+        "use strict";
+        window.onload = function what() {
+            let x = document.getElementById("main");
+            x.innerHTML = "Internal JavaScript";
```
```suggestion
            x.textContent = "Internal JavaScript";
```

Creatoon commented on JavaScript/Topics/3.Web_Browser/JS/script.js:
```diff
@@ -0,0 +1,5 @@
+"use strict";
+window.onload = function what() {
+    let x = document.getElementById("main");
+    x.innerHTML = "External JavaScript";
```
```suggestion
    x.textContent = "External JavaScript";
```

Creatoon commented on JavaScript/Topics/3.Web_Browser/README.md:
```diff
@@ -1 +1,183 @@
+    <script>
+        "use strict";
+        let x = document.getElementById("main");
+        x.innerHTML = "Internal JavaScript";
```
```suggestion
        x.textContent = "Internal JavaScript";
```

Creatoon replied:
Correct at all places where there is innerHTML and you are passing string

subhendudash02 commented:
@Creatoon I have resolved the comments and corrected "innerHTML" function.
